src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String EXCEPTION_MESSAGE = "This index is out of bound ";
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    private void checkPositionIndex(int index) {
+        if (!(index >= 0 && index <= size)) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE + index);
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!(index >= 0 && index < size)) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE + index);
+        }
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    void linkBefore(T e, Node<T> succ) {
+        Node<T> pred = succ.prev;
+        Node<T> newNode = new Node<>(pred, e, succ);
+        succ.prev = newNode;
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    void linkLast(T e) {
+        final Node<T> l = last;
+        final Node<T> newNode = new Node<>(l, e, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    T unlink(Node<T> x) {
+        Node<T> next = x.next;
+        Node<T> prev = x.prev;
+        final T element = x.item;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+        x.item = null;
+        size--;
+        return element;
+    }
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] a = list.toArray();
+        if (a.length == 0) {
+            return;
+        }
+        for (Object object : a) {
+            linkLast((T) object);
+        }
```

Comment:
We can use foreach with List<T> without creating Object[] array, and also we don't need to check for empty list, because if it is empty, cycle won't start